File: Leetcode 0449 Serialize and Deserialize Binary Search Tree [medium] [tree]/Serialize and Deserialize BST [dfs].js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * Encodes a tree to a single string.
 *
 * @param {TreeNode} root
 * @return {string}
 */
var serialize = function(root) {
    if (root === null) {
        return "";
    }
    
    let result = [""];
    dfs(root, result);
    return result[0].toString();
}; 

let dfs = (root, result) => {
    if (root === null) {
        return;
    }
    
    result[0] += root.val + ",";
    dfs(root.left, result);
    dfs(root.right, result);
}

/**
 * Decodes your encoded data to tree.
 *
 * @param {string} data
 * @return {TreeNode}
 */
var deserialize = function(data) {
    if (data === "") {
        return null;
    }
    
    let result = data.split(",");
    
    // !!!End position should be result.length - 2;
    return constructBST(result, 0, result.length - 2);
};

let constructBST = (result, start, end) => {
    if (start > end) {
        return null;
    }
    
    let cur = new TreeNode(parseInt(result[start]));
    let index;
    for (index = start; index <= end; index++) {
        if (parseInt(result[index]) > parseInt(result[start])) {
            break;
        }
    }
    
    cur.left = constructBST(result, start + 1, index - 1);
    cur.right = constructBST(result, index, end);
    
    return cur;
}

/**
 * Your functions will be called as such:
 * deserialize(serialize(root));
 */

File: Leetcode 0449 Serialize and Deserialize Binary Search Tree [medium] [tree]/Serialize and Deserialize BST [dfs].java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return "";
        }
        
        StringBuilder result = new StringBuilder();
        dfs(root, result);
        return result.toString();
    }
    
    private void dfs(TreeNode root, StringBuilder result) {
        if (root == null) {
            return;
        }
        
        result.append(root.val + ",");
        dfs(root.left, result);
        dfs(root.right, result);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        if (data.equals("")) {
            return null;
        }
        
        String[] result = data.split(",");
        return constructBST(result, 0, result.length - 1);
    }
    
    private TreeNode constructBST(String[] result, int start, int end) {
        if (start > end) {
            return null;
        }
        
        TreeNode cur = new TreeNode(Integer.parseInt(result[start]));
        int index;
        for (index = start; index <= end; index++) {
            if (Integer.parseInt(result[index]) > Integer.parseInt(result[start])) {
                break;
            }
        }
        
        cur.left = constructBST(result, start + 1, index - 1);
        cur.right = constructBST(result, index, end);
        
        return cur;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec ser = new Codec();
// Codec deser = new Codec();
// String tree = ser.serialize(root);
// TreeNode ans = deser.deserialize(tree);
// return ans;